File: mobile/ios/packages/town-crier-data/Sources/API/DotNetTimeParser.swift
import Foundation

/// Parses timestamp strings emitted by the Go backend's `platform.DotNetTime`
/// format (`2006-01-02T15:04:05.9999999-07:00`).
///
/// Fractional seconds appear on the wire **only** when the sub-second part is
/// non-zero, otherwise they are omitted entirely:
///
/// - `2026-07-20T14:23:45.6789123+00:00` — fractional (the common case, since
///   expiries derive from `time.Now()`).
/// - `2026-06-12T09:30:00+00:00` — whole second, no fractional part.
/// - `2026-06-12T09:30:00Z` — `Z` zero-offset spelling.
///
/// A single `ISO8601DateFormatter` cannot cover all three: `.withFractionalSeconds`
/// rejects whole-second strings, and the default `.withInternetDateTime` rejects
/// fractional ones. So we try fractional first, then fall back to plain. Both
/// option sets include `.withTimeZone`, which accepts the `Z` and `+hh:mm` forms.
///
/// Two `ISO8601DateFormatter` instances are cached. They are configured once and
/// never mutated afterwards, so reusing them avoids per-call allocation. The type
/// is not `Sendable`, but `ISO8601DateFormatter` is documented as safe for
/// concurrent use of its parsing methods once configured, and `formatOptions` is
/// never touched after init here — hence `nonisolated(unsafe)`.
enum DotNetTimeParser {
  private nonisolated(unsafe) static let fractional: ISO8601DateFormatter = {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter
  }()

  private nonisolated(unsafe) static let plain: ISO8601DateFormatter = {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime]
    return formatter
  }()

  /// Returns the parsed `Date`, or `nil` only if BOTH the fractional and the
  /// plain parse fail (i.e. genuinely unparseable input).
  static func date(from string: String) -> Date? {
    fractional.date(from: string) ?? plain.date(from: string)
  }
}

File: mobile/android/data/src/main/kotlin/uk/towncrierapp/data/api/DotNetTimeParser.kt
package uk.towncrierapp.data.api

import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * Parses timestamp strings emitted by the Go backend's `platform.DotNetTime`
 * format (`2006-01-02T15:04:05.9999999-07:00`) — one shared parser applied
 * per-field by DTO mappers, never a global kotlinx.serialization date
 * strategy (epic #770 "API contract essentials"; port of iOS
 * `DotNetTimeParser`).
 *
 * Fractional seconds appear on the wire **only** when the sub-second part is
 * non-zero, and the offset is spelled either `Z` or `±hh:mm`:
 *
 * - `2026-07-20T14:23:45.6789123+00:00` — fractional, numeric offset.
 * - `2026-06-12T09:30:00+00:00` — whole second, numeric offset.
 * - `2026-06-12T09:30:00Z` — whole second, `Z` offset.
 * - `2026-07-20T14:23:45.5Z` — fractional, `Z` offset.
 *
 * [DateTimeFormatter.ISO_OFFSET_DATE_TIME] already accepts every one of
 * these (0-9 optional fractional digits, `Z` or `±hh:mm`), so there is no
 * need to try multiple formatters like the iOS `ISO8601DateFormatter` port
 * does — `OffsetDateTime.parse` with the default formatter covers the whole
 * matrix in one call.
 */
public object DotNetTimeParser {
    /** Returns the parsed instant, or `null` on genuinely unparseable input — never throws. */
    public fun parse(value: String): OffsetDateTime? =
        try {
            OffsetDateTime.parse(value, DateTimeFormatter.ISO_OFFSET_DATE_TIME)
        } catch (e: DateTimeParseException) {
            null
        }

    /** Parses a bare `yyyy-MM-dd` date-only field (e.g. `startDate`, `decidedDate`) — a separate, simpler format. */
    public fun parseDate(value: String): LocalDate? =
        try {
            LocalDate.parse(value, DateTimeFormatter.ISO_LOCAL_DATE)
        } catch (e: DateTimeParseException) {
            null
        }
}
